daemon/logs.go

```diff
@@ -125,38 +126,53 @@ func (daemon *Daemon) ContainerLogs(name string, config *ContainerLogsConfig) er
 			}
 		}
 	}
+
 	if config.Follow && container.IsRunning() {
-		errors := make(chan error, 2)
-		wg := sync.WaitGroup{}
+		var (
+			stdoutErr, stderrErr   = make(chan error), make(chan error)
+			stdoutPipe, stderrPipe io.ReadCloser
+			err                    error
+		)
 
 		if config.UseStdout {
-			wg.Add(1)
-			stdoutPipe := container.StdoutLogPipe()
+			stdoutPipe = container.StdoutLogPipe()
 			defer stdoutPipe.Close()
 			go func() {
-				errors <- jsonlog.WriteLog(stdoutPipe, outStream, format)
-				wg.Done()
+				logrus.Debug("logs: stdout stream begin")
+				stdoutErr <- jsonlog.WriteLog(stdoutPipe, outStream, format)
```

Comment:
What if error would be read from another channel and you will be locked here forever?